app/src/main/java/org/piwigo/bg/UploadService.java

```diff
@@ -143,12 +151,74 @@ protected void onUploadStarted(ImageUploadQueue<UploadAction> imageUploadQueue)
             Log.e("UploadService", "IOException", e.getCause());
             content = new byte[0];
         }
-        filePart = MultipartBody.Part.createFormData("file", uploadAction.getFileName(), RequestBody.create(MediaType.parse("image/*"), content));
+
+        byte[][] splittedContent = splitArray(content, chunkSize);
+        for (int i = 0; i < splittedContent.length; i++) {
+            fileParts[i] = MultipartBody.Part.createFormData("file", uploadAction.getFileName(), RequestBody.create(MediaType.parse("image/*"), splittedContent[i]));
+        }
         restService = restServiceFactory.createForAccount(userManager.getActiveAccount().getValue());
         requestBodies = createUploadRequest(uploadAction.getFileName(), getPhotoName(uploadAction.getFileName()), userManager.getActiveAccount().getValue());
 
+        //Promise building
+        promise.setFileParts(fileParts);
+        promise.setRequestBodies(requestBodies);
+        promise.setCatId(uploadAction.getUploadData().getCategoryId());
+        promise.setCurrentChunk(0);
+        promise.setTotalChunks(expectedChunks);
         if (uploadAction.getUploadData().getCategoryId() > 0)
-            callUploadResponse(imageUploadQueue, restService, requestBodies, uploadAction.getUploadData().getCategoryId(), filePart);
+            callChunkedUploadResponse(imageUploadQueue, restService, promise);
+    }
+
+    /**
+     * Pick a chunk size depending of the network mode (if on wifi, 1MB, 256KB for mobile data)
+     * @return chunkSize (1MB, 256KB for mobile data or anything else)
+     */
+    public int getChunkSize()
+    {
+        int networkType = NetworkHelper.INSTANCE.getNetworkType(getApplicationContext());
+
+        if (networkType == ConnectivityManager.TYPE_WIFI) {
+            return (1024 * 1024); // 1MB
+        }
+        else {
+            return (256 * 1024); // 256KB
+        }
+    }
+
+    /**
+     * Get the file size from the given file URI (getting as a File does not work because of Android's way to manage temp content)
+     *
+     * @param fileUri (file path on the device)
+     * @return fileSize (as long)
+     */
+    public long getFileSize(Uri fileUri) {
+        Cursor cursor = getContentResolver().query(fileUri, null, null, null, null);
+        long size = 0;
+
+        if (cursor != null) {
+            cursor.moveToFirst();
+            size = cursor.getLong(cursor.getColumnIndex(OpenableColumns.SIZE));
+            cursor.close();
+        }
+        return size;
+    }
+
+    /**
+     * Split the file in parts to send multiple Multipart.Parts
+     *
+     * @param source    byte array containing the whole image
+     * @param chunkSize size of the chunks (1024 * 1024 = 1MB for instance)
+     * @return an array of x chunks containing chunks of source
+     */
+    public static byte[][] splitArray(byte[] source, int chunkSize) {
+        byte[][] newArray = new byte[(int) Math.ceil(source.length / (double) chunkSize)][chunkSize];
```

Comment:
Math.ceil here and "(int)(tmpFileSize / chunkSize) + 1" in onChunkedUploadStarted? Theoretically both are correct, while I have a slight preference for the integer approach I at least would request to use the same approach in both cases.

Reply:
I'll take care of this, thank you for reporting this

Reply:
not a big deal for sure. But as I recognized I "had to" comment :-) And yes, consistency is not a bad thing...

Reply:
Done in 8fe2dda